### PR TITLE
Itm 421 validator updates

### PR DESCRIPTION
**Pull request by Anchita14**
I implemented a map to keep track of all scene characters and update/remove them according to the persist_characters flag and removed_characters. I created the validate_persist_characters and update_characters methods to manage character persistence accurately. Additionally, I enhanced the scene validation logic to ensure characters marked as removed are not present in both the scene's state and action mappings.

### CODE REVIEW

**Review by nextcen-dgemoets — COMMENTED** [review]
The code assumes that the scenes are in order and don't branch.  I don't know that there's much we can do about that, other than to tweak the error messages to say "might" (see comments).

This ticket originally was going to address how `persist_characters` broke all of the character matching dependencies, but that isn't addressed in this PR.  Maybe we should move that to another ticket?  Otherwise, the approach we discussed in Teams is:

1. Check for `persist_characters` anywhere in the file first, and if it's never true, then use the character matching code we already have.
2. Build a list of all character ids anywhere in the scenario, and if they specify something that doesn't match any, then flag it as an error (increment `self.invalid_values`).
3. Check the scenario-level characters and the current scenes characters.  If an id is used that doesn't match any of them, then emit an Informational saying there might be an issue (increment `self.warning_count`).

Either in this ticket or the next, we should add a character matching rule for `"scenes[].removed_characters[]"` so that we flag if they specify an invalid character id in the `removed_characters` list.

**Comment by nextcen-dgemoets** [comment]
Oh I forgot to mention-- I think these changes should have been in a new feature branch off of `main`, e.g., `itm-466-persist-characters`, not building off of the old (merged) branch, `itm-421-validator-updates`.  I don't think we have to do anything about it (unless @kaitlyn-sharo thinks it should be fixed), but normally we create a new feature branch for each ticket.

**Comment by kaitlyn-sharo** [comment]
Yes, this should have addressed the approach you mentioned, Darren, and it should have been in a new branch

**Comment by nextcen-dgemoets** [comment]
> Yes, this should have addressed the approach you mentioned, Darren, and it should have been in a new branch

Do you want this moved to a new branch (branched off of `main`), or just keep it like this and use a new branch next time?
And do you want the character matching dependencies addressed here, or moved to another ticket?

**Comment by kaitlyn-sharo** [comment]
I think it should be addressed here. That was part of the ticket, right? If it was in the ticket, definitely address it here. I suppose this time it can stay in the branch, because it doesn't matter too much

File: persist_test.yaml [hunk 1]
```diff
@@ -0,0 +1,359 @@
+id: sample-1
```
[the remark this answers is not in the document]
Comment (kaitlyn-sharo) [reply persist_test.yaml:1]: I agree. When I said to add some test files, I meant add something specific for testing this code, like showing the correct errors showed up when things were wrong and that no errors showed up with a few different types of valid scenes.

File: validator.py [hunk 2]
```diff
@@ -221,8 +221,74 @@ def validate_one_level(self, level_name, to_validate, type_obj, required, api_ya
...
+        
+    def validate_persist_characters(self):
+        '''
+        Ensure that scene characters are valid along with the persist_characters flag.
+        '''
+        data = copy.deepcopy(self.loaded_yaml)
+        scenes = data['scenes']
+        curr_chars = data['state']['characters']
+
+        # Determine the first scene, either from 'first_scene' or the first in the scenes list
```
[the remark this answers is not in the document]
Comment (kaitlyn-sharo) [reply validator.py:235]: Totally agree. That one is more on me, I think I added the first 3 in my index-to-id branch, but @Anchita14 shouldn't be hard to do a quick find and replace here

File: validator.py [hunk 3]
```diff
@@ -221,8 +221,74 @@ def validate_one_level(self, level_name, to_validate, type_obj, required, api_ya
...
+                elif persist == False:
+                    # If not, replace the current characters with the scene's characters
+                    curr_chars = scene.get('state', {}).get('characters', [])
+
+                # Check if any characters in the current characters list are in the removed characters list
+                for char in curr_chars:
+                    char_id = char['id']
+                    if char_id in removed_chars:
+                        self.warning_count += 1
+                        self.logger.log(LogLevel.INFO, f"Character with ID {char_id} is in 'removed_characters' but is still present in the scene's state.")
```
[the remark this answers is not in the document]
Comment (kaitlyn-sharo) [reply validator.py:264]: Yes, I like the addition of "might" here and in the other location

**Comment by kaitlyn-sharo** [comment]
Updated character_matching function to handle persist_characters:
- if it is the first scene, or if the 'characters' properties exists in the scene, character mapping is handled the same as before. Look for all the locations where we might find a character id and make sure that id that's found matches a character id explicitly defined in that scene (or, in the case of the first scene, the scenario)
- if persist_characters is true, see if the character ids defined in the scene are defined anywhere in the yaml. If so, pass. Otherwise, error
- if a character has been removed at any point in time (and persist_characters is true), offer a warning to ensure that the character will be available in that scene

Logging has also been updated to define "errors" and "possible errors" instead of "warnings" and "info"

In the future, we will be investigating LOE for following scene branches to get a better idea of real errors here, instead of flagging every removed character and allowing every character that exists at any point in the scenario

**Comment by nextcen-dgemoets** [comment]
Some thoughts: 
  - So, does this mean we should remove `validate_persist_characters` and `update_characters`?
  - "if it is the first scene, or if the 'characters' properties exists in the scene, character mapping is handled the same as before"
    - this only works if persist_characters is false/missing, right?  The scene could refer to a persisted character.
  - Warnings are in purple, and warnings count is purple, but errors are in yellow and error count is red; summary is red and green
    - We could put warnings in yellow (in both places) and errors in red (in both places); then put summary in green.  Thoughts?  I may be missing the initial intent of the color-coding.
  - typo: locataion
  - In `character_matching`, could you rename `first_scene` to `first_scene_id`?  Seems a little clearer / more accurate.
  - "Character ID <id> appears in 'scenes[1].action_mapping[1].character_id', but is removed at some point during the scenario. Make sure that this character will not be removed before this scene"
    - Extremely minor, but could you change "will not be removed" to "is not removed"?
 - Should we use `determine_first_scene` in `scenes_with_state`?
   - It also looks like we could combine `scenes_with_state` and `check_first_scene`, but that's arguably beyond the scope of this ticket.

**Comment by kaitlyn-sharo** [comment]
> Some thoughts:
> 
> * So, does this mean we should remove `validate_persist_characters` and `update_characters`?
I don't think so? Anchita wrote those functions, I was just doing the character_matching part. Do they overlap? I'll have to look...

> * "if it is the first scene, or if the 'characters' properties exists in the scene, character mapping is handled the same as before"
>   
>   * this only works if persist_characters is false/missing, right?  The scene could refer to a persisted character.
Can 'characters' and 'persist_characters' coexist?

> * Warnings are in purple, and warnings count is purple, but errors are in yellow and error count is red; summary is red and green
>   
>   * We could put warnings in yellow (in both places) and errors in red (in both places); then put summary in green.  Thoughts?  I may be missing the initial intent of the color-coding.
I really want to avoid errors in red because that is a LOT of red and I find it hard to read. Colors don't even come up the same on everyone's terminal. The intent of the color-coding is so that I can have a little fun :)
I would be willing to put the end warnings as red or yellow if count > 0, but that's about it

> * typo: locataion
Okay, but where?

> * In `character_matching`, could you rename `first_scene` to `first_scene_id`?  Seems a little clearer / more accurate.
Definitely. 

> * "Character ID  appears in 'scenes[1].action_mapping[1].character_id', but is removed at some point during the scenario. Make sure that this character will not be removed before this scene"
>   
>   * Extremely minor, but could you change "will not be removed" to "is not removed"?
My thought was that it is possible for it to be removed, so we want to make sure it cannot be removed, but I think either one gets the same point across, so I'll change it

> * Should we use `determine_first_scene` in `scenes_with_state`?
>   
>   * It also looks like we could combine `scenes_with_state` and `check_first_scene`, but that's arguably beyond the scope of this ticket.
I'll look into these last two

**Comment by nextcen-dgemoets** [comment]
> > Some thoughts:
> > 
> > * So, does this mean we should remove `validate_persist_characters` and `update_characters`?
> I don't think so? Anchita wrote those functions, I was just doing the character_matching part. Do they overlap? I'll have to look...

Well if nothing else, `validate_persist_characters` is no longer invoked.  And I *believe* it is covered by the `removed_characters` character matching rule.
 
> > * "if it is the first scene, or if the 'characters' properties exists in the scene, character mapping is handled the same as before"
> >   
> >   * this only works if persist_characters is false/missing, right?  The scene could refer to a persisted character.
> 
> Can 'characters' and 'persist_characters' coexist?

Yes, if `persist_characters` is true, then any characters you specify in `persist_characters` are added to the scene's characters (replacing any characters with the same `id`.  I'm not sure that's well-documented (for users, let alone devs, so I'll look to address that).

> > * Warnings are in purple, and warnings count is purple, but errors are in yellow and error count is red; summary is red and green
> >   
> >   * We could put warnings in yellow (in both places) and errors in red (in both places); then put summary in green.  Thoughts?  I may be missing the initial intent of the color-coding.
> 
> I really want to avoid errors in red because that is a LOT of red and I find it hard to read. Colors don't even come up the same on everyone's terminal. The intent of the color-coding is so that I can have a little fun :) I would be willing to put the end warnings as red or yellow if count > 0, but that's about it

I agree that a lot of red is hard to read.

**Review by nextcen-dgemoets — APPROVED** [review]
A little more discussion in Teams, and we are good to go!

**Pull request merged**